### ISSUE TITLE
Passing in shard_num metadata to delta/snapshot files 

### ISSUE THREAD

**Issue by ahmad-masad** [privacysandbox/protected-auction-key-value-service#33]
Hi! 

Our team has been working on incorporating sharding in to the key value store. We are currently struggling with specifying the shard number a delta file/snapshot belongs to. We are aware of the existence of the shard_num metadata field upon reviewing the sharding documents and assume that we would use the data_cli tool to pass in this field. At first, it seemed that we would need to pass it in via the csv file headers, but it appears that the data_cli tool expects specific headers for the default mutation record type and those headers do not include a shard specification. Could you please point us towards how to pass in this metadata? 

On a slightly separate note, we noticed other metadata fields such as number of logical shards and number of physical shards as well as the ability to pass in a sharding config via the data_cli tool. Our understanding is that the sharding config is used to map logical shard numbers to physical shard numbers? Is it expected that the mapping of logical to physical shards is taken care of from our side? Do we need to specify these other metadata fields such as number of logical and physical shards to be able to use sharding or do our responsibilities end at what was mentioned in the sharding documents? 
(By responsibilities, I am specifically referencing:
Standard path- Grouping the key/values that belong to the same shard in a unique delta file by using the provided sharding function, passing in that shard_num metadata field
Realtime path- Passing in the shard number to the SNS message)
Thank you!

**Comment by peiwenhu** [comment]
Hi,

The shard number in the metadata is an optimization (probably required for production but not as much for experimentation). It allows the server to skip a whole file if the records are for a different shard but if it isn't present, the server will still filter out records in other shards. For your purpose I suppose you could first leave this out just to move on to test the rest.

For this problem itself, to clarify, the data_cli tool is not a requirement but a convenient utility. We expected users to incorporate the libraries ([Riegeli](https://github.com/google/riegeli) for the file&metadata, [Flatbuffers](https://flatbuffers.dev/) for the actual records, and [some of our own libraries](https://github.com/privacysandbox/fledge-key-value-service/tree/release-0.14/public/data_loading) for convenience) into their own data generation system for actual production integration. Since the shard_num is an optimization we may not have the same level of support in the data_cli yet. We're interested to know if this matches your expectation. If you plan to use the data_cli more extensively we can prioritize adding support over there. I imagine one pain point may be the lack of support of those libraries for other languages. It was a tradeoff we had to make.

The sharding config is for the logical/physical shards yes. The logical/physical shard is still experimental and not yet ready (logical to physical shard mapping is intended to be configred from your side). It is intended to support easier resharding. We could explain more about resharding if you're interested. But TL;DR is for now you can ignore these fields. You don't need them to use sharding. It's just we don't have a smooth resharding experience yet. You'd have to bring up a different kv fleet with new shards, redirect the traffic and shut down the old fleet.

TL;DR Right now I think the minimum responsibility is to set shard_num in the terraform and bring up the instance with that. The data generation side does not have requirement for any changes.

We'll update the sharding document to explain it more clearly.

**Comment by lx3-g** [comment]
Here's an example how to set the shard_num metadata field -- https://github.com/privacysandbox/fledge-key-value-service/blob/252d361c7a3b291f50ffbf36d86fc4405af6a147/tools/serving_data_generator/test_serving_data_generator.cc#L118

**Comment by ahmad-masad** [comment]
Understood, we are looking in to moving away from relying as heavily on the data_cli tool and will update you if we encounter any pain points

**Comment by peiwenhu** [comment]
Got it. Happy to help in any way and we are also very interested in your use case to see what we can prioritize on our side. If you're interested please feel free to contact privacy-sandbox-trusted-kv-server-support@google.com

**Comment by ahmad-masad** [comment]
After looking in to some of the suggested approaches, the team wanted to share the following feedback:

Flatbuffers seem to be supported in a wide set of languages, with at least some kind of community around each, that's ok-ish. But Riegeli looks abandoned and without wide adoption, requiring extra efforts to figure out how to do things (but also to write the different extensions in the languages of our different systems in Java, Rust, Python, to name a few). Can Google reconsider switching to a more supported (i.e: bigger community) format for the files? Or maybe make data_cli a "formal" part (or player) in the kv server tooling? (i.e: add support to write this field, and perhaps all the metadata fields).

In general, we still feel that the data_cli tool is the quickest and most language-agnostic way of writing these files.

**Comment by peiwenhu** [comment]
The shard_num support is added in https://github.com/privacysandbox/fledge-key-value-service/commit/95c0b96effcd97d037eb08521b32158fac5489d1
This is available in the `main` branch hosting the latest code. The next coming release will also have this.

While having a bigger community is nice to have to get help, we have to make our decision based on the actual production performance given how challenging it is to implement a real time server on the critical path of auction inside TEE. Getting help and figuring out how to use a tool is an one-time investment while keeping the system running at the optimal/acceptable performance has far greater impact over the long term. Unfortunately we have not found the perfect choice for everything so It's a bet we have to make.

Closing this ticket as the commit above fixes the original request.
